3-0-spring-framework/3-3-0-enable-string-trimming/src/main/java/com/bobocode/TrimmedAnnotationBeanPostProcessor.java

```diff
@@ -1,21 +1,106 @@
  * {@link Trimmed}. For example if there is a string " Java   " as an input parameter it has to be automatically trimmed to "Java"
  * if parameter is marked with {@link Trimmed} annotation.
  * <p>
- *
+ * <p>
  * Note! This bean is not marked as a {@link Component} to avoid automatic scanning, instead it should be created in
  * {@link StringTrimmingConfiguration} class which can be imported to a {@link Configuration} class by annotation
  * {@link EnableStringTrimming}
  */
-public class TrimmedAnnotationBeanPostProcessor {
-//todo: Implement TrimmedAnnotationBeanPostProcessor according to javadoc
+public class TrimmedAnnotationBeanPostProcessor implements BeanPostProcessor {
+
+    private final Map<String, Object> beansToBeProxied = new HashMap<>();
+    private final Map<String, MethodInterceptor[]> beanInterceptors = new HashMap<>();
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        if (containsParametersAnnotatedWithTrimmed(bean)) {
+            beansToBeProxied.put(beanName, bean);
+            beanInterceptors.put(beanName, new MethodInterceptor[]{provideTrimmingInterceptor()});
+        }
+        return BeanPostProcessor.super.postProcessBeforeInitialization(bean, beanName);
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        Object toProxy = beansToBeProxied.get(beanName);
+        if (toProxy != null) {
+            Enhancer enhancer = new Enhancer();
+            enhancer.setSuperclass(toProxy.getClass());
+            setInterceptorsIfNeeded(enhancer, bean, beanName);
+            return enhancer.create();
+        }
+        return bean;
+    }
+
+    private MethodInterceptor provideTrimmingInterceptor() {
+        return (beanInstance, method, args, methodProxy) ->
+                methodProxy.invokeSuper(beanInstance, processParameters(method, args));
+    }
+
+    private Object[] processParameters(Method method, Object[] args) {
+        Parameter[] parameters = method.getParameters();
+        for (int i = 0; i < parameters.length; i++) {
+            if (parameters[i].isAnnotationPresent(Trimmed.class)) {
+                args[i] = trimMarkedString(args[i]);
+            }
+        }
+        return args;
+    }
+
+    private Object trimMarkedString(Object arg) {
+        if (arg instanceof String string) {
+            return StringUtils.hasLength(string) ? string.trim() : arg;
+        }
+        return arg;
+    }
+
+    private boolean containsParametersAnnotatedWithTrimmed(Object bean) {
+        return Arrays.stream(bean.getClass().getDeclaredMethods()).flatMap(m -> Stream.of(m.getParameters()))
+                .anyMatch(p -> p.isAnnotationPresent(Trimmed.class));
+    }
+
+    private void setInterceptorsIfNeeded(Enhancer enhancer, Object bean, String beanName) {
+        MethodInterceptor[] methodInterceptors = beanInterceptors.get(beanName);
```

Comment:
@Stanislav-Zabramnyi to me **this second map `beanInterceptors` is redundant**. The only thing it's storing is an interceptor  created by `provideTrimmingInterceptor`. So we should get rid of that map, and just call `provideTrimmingInterceptor` when we need it.